packages/ubuntu_wsl_setup/lib/pages/applying_changes/applying_changes_page.dart

```diff
@@ -0,0 +1,67 @@
+import 'package:flutter/material.dart';
+import 'package:provider/provider.dart';
+import 'package:subiquity_client/subiquity_client.dart';
+import 'package:ubuntu_wizard/constants.dart';
+import 'package:ubuntu_wizard/services.dart';
+import 'package:ubuntu_wizard/widgets.dart';
+
+import '../../l10n/app_localizations.dart';
+import 'applying_changes_model.dart';
+
+/// WSL Advanced setup page.
+///
+/// See also:
+/// * [AdvancedSetupModel]
+class ApplyingChangesPage extends StatefulWidget {
+  /// Use [create] instead.
+  @visibleForTesting
+  const ApplyingChangesPage({
+    Key? key,
+  }) : super(key: key);
+
+  /// Creates an instance with [AdvancedSetupModel].
+  static Widget create(BuildContext context) {
+    final monitor = getService<SubiquityStatusMonitor>();
+    return Provider(
+      create: (_) => ApplyingChangesModel(monitor),
+      child: const ApplyingChangesPage(),
+    );
+  }
+
+  @override
+  State<ApplyingChangesPage> createState() => _ApplyingChangesPageState();
+}
+
+class _ApplyingChangesPageState extends State<ApplyingChangesPage> {
+  bool previousData = true;
+  @override
+  Widget build(BuildContext context) {
+    final model = Provider.of<ApplyingChangesModel>(context, listen: false);
+    final lang = AppLocalizations.of(context);
+    return WizardPage(
+      title: Text(lang.setupCompleteTitle),
+      content: StreamBuilder<bool>(
+          stream: model.isInstalling(),
+          builder: (context, snapshot) {
+            if (snapshot.data == false && previousData == true) {
```

Comment:
this smells like doing business logic in the view :)

Reply:
Well, you're right. And you might be able to help with understand why this was necessary.

Without this, I get an unhandled exception at the end of the wizard due the last route calling `Wizard.next()` when I click too fast at the exit button. The same behavior can be more easily observed by running the integration tests.

I thought that could be some sort of race condition caused by the Streambuilder scheduling the post frame callback more than once. Lift this logic to the model should be able to avoid the stream to issue `false` more than once, so the post frame callback should never be scheduled more than once. But that's not what happen. Making this widget stateless makes things worse. So apparently the only way to avoid the StreamBuilder to schedule the post frame callback more than once is by keeping some state in the widget.

Even if I do this in the ApplyingChanges in the model, keep the widget stateful but remove the `previousData` from the comparison in the if-statement, I still get the referred exceptions.

Model:
```diff
-  Stream<bool> isInstalling() => _isInstalling;
+  bool previous = true;
+
+  // Returns false only once, after the server first leaving the
+  // [InstallingState] (according to that extension definition).
+  // This mitigates the risk of Wizard.next() being called by the last page due
+  // a fast transition on this stream.
+  Stream<bool> isInstalling() => _isInstalling.map((event) {
+        if (event == false && previous == true) {
+          previous = false;
+          return false;
+        }
+        return true;
+      });

```

View:
```diff
-          if (snapshot.data == false && previousData == true) {
+          if (snapshot.data == false) {
```

The kind of stack trace I got:


```
flutter: DEBUG subiquity_client: GET http://localhost/identity
flutter: DEBUG subiquity_client: ==> identity() {"realname": "u", "username": "u", "crypted_password": "", "hostname": ""}
flutter: INFO system_shutdown: request reboot (immediate: false)
flutter: INFO window: request window close
starting server
flutter: INFO window: window closed
flutter: INFO subiquity_client: Closing socket /home/cnihelton/Dev/Flutter_UDI/main/packages/ubuntu_wsl_setup/../subiquity_client/subiquity/.subiquity/socket
flutter: DEBUG [CN]: Scheduling post frame callback
flutter: DEBUG [CN]: Running the Post frame callback
flutter: ERROR ubuntu_wsl_setup: Unhandled exception
        _AssertionError: 'package:wizard_router/src/wizard.dart': Failed assertion: line 324 pos 14: 'index < widget._routes.length - 1': `Wizard.next()` called from the last route /setup-complete.
\#0      _AssertionError._doThrowNew (dart:core-patch/errors_patch.dart:51:61)
\#1      _AssertionError._throwNew (dart:core-patch/errors_patch.dart:40:5)
\#2      WizardScopeState.next.nextRoute (package:wizard_router/src/wizard.dart:324:14)
\#3      WizardScopeState.next (package:wizard_router/src/wizard.dart:330:25)
\#4      ApplyingChangesPage.build.<anonymous closure>.<anonymous closure> (package:ubuntu_wsl_setup/pages/applying_changes/applying_changes_page.dart:47:36)
\#5      SchedulerBinding._invokeFrameCallback (package:flutter/src/scheduler/binding.dart:1144:15)
\#6      SchedulerBinding.handleDrawFrame (package:flutter/src/scheduler/binding.dart:1089:9)
\#7      SchedulerBinding._handleDrawFrame (package:flutter/src/scheduler/binding.dart:995:5)
\#8      _rootRun (dart:async/zone.dart:1426:13)
\#9      _CustomZone.run (dart:async/zone.dart:1328:19)
\#10     _CustomZone.runGuarded (dart:async/zone.dart:1236:7)
\#11     _invoke (dart:ui/hooks.dart:151:10)
\#12     PlatformDispatcher._drawFrame (dart:ui/platform_dispatcher.dart:308:5)
\#13     _drawFrame (dart:ui/hooks.dart:115:31)

Lost connection to device.
```

Would you have a better idea other than attempting to replace the StramBuilder by some hand-written similar widget?

Reply:
In a way, the view model is not doing its job but merely exposing a raw stream and that way pushing the processing logic to the view. What if we'd process the stream in the view model instead, and drop the stream from the view?

```dart
class ApplyingChangesModel extends ChangeNotifier {
  ApplyingChangesModel(this._monitor);

  final SubiquityStatusMonitor _monitor;
  StreamSubscription<ApplicationStatus?>? _sub;

  bool get isInstalling => _monitor.status?.state?.isInstalling == true;

  void init({required VoidCallback onDone}) {
    _sub = _monitor.onStatusChanged.listen((status) {
      if (status?.state == ApplicationState.DONE ||
          status?.state == ApplicationState.ERROR) {
        onDone();
      } else {
        notifyListeners();
      }
    });
  }

  @override
  void dispose() {
    _sub?.cancel();
    super.dispose();
  }
}
```

With this, the widget tree becomes fully declarative. In fact, `isInstalling` is no longer even needed because this view only visualizes that state.
```dart
class _ApplyingChangesPageState extends State<ApplyingChangesPage> {
  @override
  void initState() {
    super.initState();
    final model = Provider.of<ApplyingChangesModel>(context, listen: false);
    model.init(onDone: Wizard.of(context).next);
  }

  @override
  Widget build(BuildContext context) {
    final lang = AppLocalizations.of(context);
    return WizardPage(
      title: Text(lang.setupCompleteTitle),
      content: Column(
        mainAxisAlignment: MainAxisAlignment.center,
        children: [
          const CircularProgressIndicator(),
          const SizedBox(height: kContentSpacing),
          Text(lang.applyingChanges),
        ],
      ),
    );
  }
}
```

Reply:
You're absolutely right. This approach is still simple enough and prevents the issue I mentioned. I was too biased into using StreamBuilder. Thanks for the input.